agent-connection/src/main/kotlin/com/luxoft/blockchainlab/corda/hyperledger/indy/PythonRefAgentConnection.kt

```diff
@@ -57,137 +92,174 @@ class PythonRefAgentConnection : AgentConnection {
         }
     }
 
-    override fun acceptInvite(invite: String) {
-        if (invite != null && getConnectionStatus() == AgentConnectionStatus.AGENT_CONNECTED) {
-            sendJson(ReceiveInviteMessage(invite))
-            val invite = webSocket.waitForMessageOfType<InviteReceivedMessage>(MESSAGE_TYPES.INVITE_RECEIVED)
-            sendRequest(invite.key)
-            val response = webSocket.waitForMessageOfType<RequestResponseReceivedMessage>(MESSAGE_TYPES.RESPONSE_RECEIVED)
-            counterParty = IndyParty(response.their_did, invite.endpoint)
-        } else {
-            throw AgentConnectionException("AgentConnection object has wrong state")
-        }
-    }
-
-    fun sendJson(obj: Any) = webSocket.sendJson(obj)
-
-    override fun generateInvite(): String {
-        webSocket.sendJson(SendMessage(type = MESSAGE_TYPES.GENERATE_INVITE, id = Instant.now().toEpochMilli().toString()))
-        return webSocket.waitForMessageOfType<ReceiveInviteMessage>(MESSAGE_TYPES.INVITE_GENERATED).invite
-    }
-
-    override fun waitForInvitedParty() {
-        if (getConnectionStatus() == AgentConnectionStatus.AGENT_CONNECTED) {
-            webSocket.waitForMessageOfType<RequestReceivedMessage>(MESSAGE_TYPES.REQUEST_RECEIVED).also {
-                counterParty = IndyParty(it.did, it.endpoint)
-                sendJson(RequestSendResponseMessage(it.did))
+    /**
+     * Establishes a connection to remote IndyParty based on the given invite
+     */
+    override fun acceptInvite(invite: String): Single<IndyPartyConnection> {
+        return Single.create { observer ->
+            try {
+                if (getConnectionStatus() == AgentConnectionStatus.AGENT_CONNECTED) {
+                    /**
+                     * To accept the invite, first inform the agent with "receive_invite" message
+                     */
+                    sendJson(ReceiveInviteMessage(invite))
+                    val pKey = getPubkeyFromInvite(invite)
+                    /**
+                     * The agent must respond with "invite_received" message, containing the public key from invite
+                     */
+                    webSocket.receiveMessageOfType<InviteReceivedMessage>(MESSAGE_TYPES.INVITE_RECEIVED, pKey).subscribe({ invRcv ->
+                        /**
+                         * Now instruct the agent to send the connection request with "send_request" message.
+                         * The agent builds up the connection request and forwards it to the other IndyParty's endpoint,
+                         * recalling the invite by its public key
+                         */
+                        sendRequest(invRcv.key)
```

Comment:
Should we check that 
`invRcv.key == pKey`

BTW could we use pubKey instead of pKey, because it could be considered as privateKey as well as publicKey

Reply:
No we must not check that, because webSocket.receiveMessageOfType<InviteReceivedMessage>(MESSAGE_TYPES.INVITE_RECEIVED, pubKey) observes only messages coming with exactly the same pubKey. I'll fix the naming.

Reply:
done.